Make Flow topology safe on duplicate for JDBC

some flow topology insert can crash the application : 
```
org.jooq.exception.IntegrityConstraintViolationException: SQL [insert into flow_topologies ("key", "value") values ('***, cast('{"source":{"uid":"'***","tenantId":"'***","namespace":"'***","id":"'***"},"relation":"FLOW_TASK","destination":{"uid":"'***","tenantId":"'***","namespace":"'***","id":"'***"}}' as jsonb))]; Batch entry 5 insert into flow_topologies ("key", "value") values (''***', cast('{"source":{"uid":"'***","tenantId":"'***","namespace":"'***","id":"'***"},"relation":"FLOW_TASK","destination":{"uid":"'***","tenantId":"'***","namespace":"'***","id":"'***"}}' as jsonb)) was aborted: ERROR: duplicate key value violates unique constraint "flow_topologies_pkey"
```
that lead to 
```
2024-09-03 15:19:36,795 WARN  jdbc-queue-Execution_0 c.zaxxer.hikari.pool.ProxyConnection HikariPool-1 - Connection org.postgresql.jdbc.PgConnection@1bec7e3 marked as broken because of SQLSTATE(08006), ErrorCode(0)
org.postgresql.util.PSQLException: An I/O error occurred while sending to the backend.
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:398)
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:329)
	at org.postgresql.jdbc.PgConnection.executeTransactionCommand(PgConnection.java:981)
	at org.postgresql.jdbc.PgConnection.commit(PgConnection.java:1003)
	at com.zaxxer.hikari.pool.ProxyConnection.commit(ProxyConnection.java:378)
	at com.zaxxer.hikari.pool.HikariProxyConnection.commit(HikariProxyConnection.java)
	at io.micronaut.transaction.jdbc.DataSourceTransactionManager.doCommit(DataSourceTransactionManager.java:156)
	at io.micronaut.transaction.jdbc.DataSourceTransactionManager.doCommit(DataSourceTransactionManager.java:53)
	at io.micronaut.transaction.support.AbstractTransactionOperations.commitInternal(AbstractTransactionOperations.java:379)
	at io.micronaut.transaction.support.AbstractTransactionOperations.commit(AbstractTransactionOperations.java:571)
	at io.micronaut.configuration.jooq.MicronautTransactionProvider.commit(MicronautTransactionProvider.java:62)
	at org.jooq.impl.DefaultDSLContext.lambda$transactionResult0$3(DefaultDSLContext.java:534)
	at org.jooq.impl.Tools$3$1.block(Tools.java:6325)
	at java.base/java.util.concurrent.ForkJoinPool.unmanagedBlock(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool.managedBlock(Unknown Source)
	at org.jooq.impl.Tools$3.get(Tools.java:6322)
	at org.jooq.impl.DefaultDSLContext.transactionResult0(DefaultDSLContext.java:578)
	at org.jooq.impl.DefaultDSLContext.transactionResult(DefaultDSLContext.java:502)
	at org.jooq.impl.DefaultDSLContext.transaction(DefaultDSLContext.java:591)
	at io.kestra.jdbc.JooqDSLContextWrapper.lambda$transaction$1(JooqDSLContextWrapper.java:58)
	at dev.failsafe.Functions.lambda$toCtxSupplier$11(Functions.java:243)
	at dev.failsafe.Functions.lambda$get$0(Functions.java:46)
	at dev.failsafe.internal.RetryPolicyExecutor.lambda$apply$0(RetryPolicyExecutor.java:74)
	at dev.failsafe.internal.FallbackExecutor.lambda$apply$0(FallbackExecutor.java:51)
	at dev.failsafe.SyncExecutionImpl.executeSync(SyncExecutionImpl.java:187)
	at dev.failsafe.FailsafeExecutor.call(FailsafeExecutor.java:376)
	at dev.failsafe.FailsafeExecutor.get(FailsafeExecutor.java:112)
	at io.kestra.core.utils.RetryUtils$Instance.wrap(RetryUtils.java:144)
	at io.kestra.core.utils.RetryUtils$Instance.runRetryIf(RetryUtils.java:103)
	at io.kestra.jdbc.JooqDSLContextWrapper.transaction(JooqDSLContextWrapper.java:55)
	at io.kestra.jdbc.runner.JdbcQueue.produce(JdbcQueue.java:101)
	at io.kestra.jdbc.runner.JdbcQueue.emitOnly(JdbcQueue.java:116)
	at io.kestra.jdbc.runner.JdbcExecutor.toExecution(JdbcExecutor.java:764)
	at io.kestra.jdbc.runner.JdbcExecutor.executionQueue(JdbcExecutor.java:462)
	at java.base/java.util.ArrayList.forEach(Unknown Source)
	at io.kestra.jdbc.runner.JdbcQueue.lambda$receive$6(JdbcQueue.java:197)
	at io.kestra.jdbc.runner.JdbcQueue.lambda$receiveImpl$9(JdbcQueue.java:246)
	at io.kestra.jdbc.runner.JdbcQueue.lambda$poll$10(JdbcQueue.java:269)
	at io.micrometer.core.instrument.internal.TimedRunnable.run(TimedRunnable.java:49)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
Caused by: java.net.SocketException: Socket closed
	at java.base/sun.nio.ch.NioSocketImpl.endRead(Unknown Source)
	at java.base/sun.nio.ch.NioSocketImpl.implRead(Unknown Source)
	at java.base/sun.nio.ch.NioSocketImpl.read(Unknown Source)
	at java.base/sun.nio.ch.NioSocketImpl$1.read(Unknown Source)
	at java.base/java.net.Socket$SocketInputStream.read(Unknown Source)
	at org.postgresql.core.VisibleBufferedInputStream.readMore(VisibleBufferedInputStream.java:162)
	at org.postgresql.core.VisibleBufferedInputStream.ensureBytes(VisibleBufferedInputStream.java:129)
	at org.postgresql.core.VisibleBufferedInputStream.ensureBytes(VisibleBufferedInputStream.java:114)
	at org.postgresql.core.VisibleBufferedInputStream.read(VisibleBufferedInputStream.java:74)
	at org.postgresql.core.PGStream.receiveChar(PGStream.java:467)
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2166)
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:371)
	... 41 common frames omitted
2024-09-03 15:19:36,796 ERROR jdbc-queue-Execution_0 i.m.t.j.DataSourceTransactionManager Commit exception overridden by rollback exception
io.micronaut.transaction.exceptions.TransactionSystemException: Could not commit JDBC transaction
	at io.micronaut.transaction.jdbc.DataSourceTransactionManager.doCommit(DataSourceTransactionManager.java:158)
	at io.micronaut.transaction.jdbc.DataSourceTransactionManager.doCommit(DataSourceTransactionManager.java:53)
	at io.micronaut.transaction.support.AbstractTransactionOperations.commitInternal(AbstractTransactionOperations.java:379)
	at io.micronaut.transaction.support.AbstractTransactionOperations.commit(AbstractTransactionOperations.java:571)
	at io.micronaut.configuration.jooq.MicronautTransactionProvider.commit(MicronautTransactionProvider.java:62)
	at org.jooq.impl.DefaultDSLContext.lambda$transactionResult0$3(DefaultDSLContext.java:534)
	at org.jooq.impl.Tools$3$1.block(Tools.java:6325)
	at java.base/java.util.concurrent.ForkJoinPool.unmanagedBlock(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool.managedBlock(Unknown Source)
	at org.jooq.impl.Tools$3.get(Tools.java:6322)
	at org.jooq.impl.DefaultDSLContext.transactionResult0(DefaultDSLContext.java:578)
	at org.jooq.impl.DefaultDSLContext.transactionResult(DefaultDSLContext.java:502)
	at org.jooq.impl.DefaultDSLContext.transaction(DefaultDSLContext.java:591)
	at io.kestra.jdbc.JooqDSLContextWrapper.lambda$transaction$1(JooqDSLContextWrapper.java:58)
	at dev.failsafe.Functions.lambda$toCtxSupplier$11(Functions.java:243)
	at dev.failsafe.Functions.lambda$get$0(Functions.java:46)
	at dev.failsafe.internal.RetryPolicyExecutor.lambda$apply$0(RetryPolicyExecutor.java:74)
	at dev.failsafe.internal.FallbackExecutor.lambda$apply$0(FallbackExecutor.java:51)
	at dev.failsafe.SyncExecutionImpl.executeSync(SyncExecutionImpl.java:187)
	at dev.failsafe.FailsafeExecutor.call(FailsafeExecutor.java:376)
	at dev.failsafe.FailsafeExecutor.get(FailsafeExecutor.java:112)
	at io.kestra.core.utils.RetryUtils$Instance.wrap(RetryUtils.java:144)
	at io.kestra.core.utils.RetryUtils$Instance.runRetryIf(RetryUtils.java:103)
	at io.kestra.jdbc.JooqDSLContextWrapper.transaction(JooqDSLContextWrapper.java:55)
	at io.kestra.jdbc.runner.JdbcQueue.produce(JdbcQueue.java:101)
	at io.kestra.jdbc.runner.JdbcQueue.emitOnly(JdbcQueue.java:116)
	at io.kestra.jdbc.runner.JdbcExecutor.toExecution(JdbcExecutor.java:764)
	at io.kestra.jdbc.runner.JdbcExecutor.executionQueue(JdbcExecutor.java:462)
	at java.base/java.util.ArrayList.forEach(Unknown Source)
	at io.kestra.jdbc.runner.JdbcQueue.lambda$receive$6(JdbcQueue.java:197)
	at io.kestra.jdbc.runner.JdbcQueue.lambda$receiveImpl$9(JdbcQueue.java:246)
	at io.kestra.jdbc.runner.JdbcQueue.lambda$poll$10(JdbcQueue.java:269)
	at io.micrometer.core.instrument.internal.TimedRunnable.run(TimedRunnable.java:49)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
Caused by: org.postgresql.util.PSQLException: An I/O error occurred while sending to the backend.
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:398)
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:329)
	at org.postgresql.jdbc.PgConnection.executeTransactionCommand(PgConnection.java:981)
	at org.postgresql.jdbc.PgConnection.commit(PgConnection.java:1003)
	at com.zaxxer.hikari.pool.ProxyConnection.commit(ProxyConnection.java:378)
	at com.zaxxer.hikari.pool.HikariProxyConnection.commit(HikariProxyConnection.java)
	at io.micronaut.transaction.jdbc.DataSourceTransactionManager.doCommit(DataSourceTransactionManager.java:156)
	... 35 common frames omitted
Caused by: java.net.SocketException: Socket closed
	at java.base/sun.nio.ch.NioSocketImpl.endRead(Unknown Source)
	at java.base/sun.nio.ch.NioSocketImpl.implRead(Unknown Source)
	at java.base/sun.nio.ch.NioSocketImpl.read(Unknown Source)
	at java.base/sun.nio.ch.NioSocketImpl$1.read(Unknown Source)
	at java.base/java.net.Socket$SocketInputStream.read(Unknown Source)
	at org.postgresql.core.VisibleBufferedInputStream.readMore(VisibleBufferedInputStream.java:162)
	at org.postgresql.core.VisibleBufferedInputStream.ensureBytes(VisibleBufferedInputStream.java:129)
	at org.postgresql.core.VisibleBufferedInputStream.ensureBytes(VisibleBufferedInputStream.java:114)
	at org.postgresql.core.VisibleBufferedInputStream.read(VisibleBufferedInputStream.java:74)
	at org.postgresql.core.PGStream.receiveChar(PGStream.java:467)
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2166)
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:371)
	... 41 common frames omitted
2024-09-03 15:19:36,799 WARN  command-shutdown io.kestra.cli.AbstractCommand Receiving shutdown ! Try to graceful exit
2024-09-03 15:19:36,799 INFO  command-shutdown i.kestra.core.contexts.KestraContext Kestra server - Shutdown initiated
2024-09-03 15:19:36,799 INFO  command-shutdown i.kestra.core.contexts.KestraContext Kestra server - Shutdown completed
```
We should: 
- Be sure to be transaction safe, the topology can be edited on multiple thread
- or use a simple `REPLACE INTO` instead `INSERT INTO` if we are not sure at 100%, it will don't change the data behind

--- a/jdbc-mysql/src/main/java/io/kestra/repository/mysql/MysqlFlowTopologyRepository.java
+++ b/jdbc-mysql/src/main/java/io/kestra/repository/mysql/MysqlFlowTopologyRepository.java
@@ -2,9 +2,13 @@ package io.kestra.repository.mysql;
 
 import io.kestra.core.models.topologies.FlowTopology;
 import io.kestra.jdbc.repository.AbstractJdbcFlowTopologyRepository;
+import io.kestra.jdbc.repository.AbstractJdbcRepository;
 import jakarta.inject.Inject;
 import jakarta.inject.Named;
 import jakarta.inject.Singleton;
+import org.jooq.DMLQuery;
+import org.jooq.DSLContext;
+import org.jooq.Record;
 
 @Singleton
 @MysqlRepositoryEnabled
@@ -12,5 +16,14 @@ public class MysqlFlowTopologyRepository extends AbstractJdbcFlowTopologyReposit
     @Inject
     public MysqlFlowTopologyRepository(@Named("flowtopologies") MysqlRepository<FlowTopology> repository) {
         super(repository);
+    }
+
+    @Override
+    protected DMLQuery<Record> buildMergeStatement(DSLContext context, FlowTopology flowTopology) {
+        return context.insertInto(this.jdbcRepository.getTable())
+            .set(AbstractJdbcRepository.field("key"), this.jdbcRepository.key(flowTopology))
+            .set(this.jdbcRepository.persistFields(flowTopology))
+            .onDuplicateKeyUpdate()
+            .set(this.jdbcRepository.persistFields(flowTopology));
     }
 }

--- a/jdbc/src/main/java/io/kestra/jdbc/repository/AbstractJdbcFlowTopologyRepository.java
+++ b/jdbc/src/main/java/io/kestra/jdbc/repository/AbstractJdbcFlowTopologyRepository.java
@@ -4,14 +4,13 @@ import io.kestra.core.models.flows.Flow;
 import io.kestra.core.models.topologies.FlowTopology;
 import io.kestra.core.repositories.FlowTopologyRepositoryInterface;
 import io.kestra.jdbc.runner.JdbcIndexerInterface;
-import jakarta.inject.Singleton;
 import org.jooq.*;
+import org.jooq.Record;
 import org.jooq.impl.DSL;
 
 import java.util.ArrayList;
 import java.util.List;
 import java.util.Map;
-import java.util.stream.Collectors;
 
 public abstract class AbstractJdbcFlowTopologyRepository extends AbstractJdbcRepository implements FlowTopologyRepositoryInterface, JdbcIndexerInterface<FlowTopology> {
     protected final io.kestra.jdbc.AbstractJdbcRepository<FlowTopology> jdbcRepository;
@@ -105,15 +104,23 @@ public abstract class AbstractJdbcFlowTopologyRepository extends AbstractJdbcRep
                     context
                         .batch(flowTopologies
                             .stream()
-                            .map(flowTopology -> context.insertInto(this.jdbcRepository.getTable())
-                                .set(AbstractJdbcRepository.field("key"), this.jdbcRepository.key(flowTopology))
-                                .set(this.jdbcRepository.persistFields(flowTopology))
-                            )
+                            .map(flowTopology -> buildMergeStatement(context, flowTopology))
                             .toList()
                         )
                         .execute();
                 }
             });
+    }
+
+    protected DMLQuery<Record> buildMergeStatement(DSLContext context, FlowTopology flowTopology) {
+        return context.mergeInto(this.jdbcRepository.getTable())
+            .using(context.selectOne())
+            .on(AbstractJdbcRepository.field("key").eq(this.jdbcRepository.key(flowTopology)))
+            .whenMatchedThenUpdate()
+            .set(this.jdbcRepository.persistFields(flowTopology))
+            .whenNotMatchedThenInsert()
+            .set(AbstractJdbcRepository.field("key"), this.jdbcRepository.key(flowTopology))
+            .set(this.jdbcRepository.persistFields(flowTopology));
     }
 
     @Override

